File: numeric/primes500/primes.cpp
#include <print>
#include <vector>
#include <generator>
#include <algorithm>
#include <ranges>

std::generator<const int> prime_sieve()
{
    co_yield 2;
    std::vector<int> cache = {2};
    int candidate = 3;
    while (true)
    {
        if (std::all_of(cache.begin(), cache.end(),
            [=](int prime) { return candidate % prime != 0; }))
        {
            co_yield candidate;
            cache.push_back(candidate);
        }

        candidate += 2;
    }
}

std::vector<int> get_primes(const int count)
{
    auto view = prime_sieve() | std::views::take(count);
    std::vector<int> primes;
    for (const int prime : view)
    {
        primes.push_back(prime);
    }
    return primes;
}

void print_primes(const std::vector<int>& primes)
{
    std::println("First Five Hundred Primes");
    for (int i = 0; i < 50; ++i)
    {
        std::println(
            "     {:04} {:04} {:04} {:04} {:04} {:04} {:04} {:04} {:04} {:04}",
            primes[i], primes[50 + i], primes[100 + i], primes[150 + i],
            primes[200 + i], primes[250 + i], primes[300 + i],
            primes[350 + i], primes[400 + i], primes[450 + i]);
    }
}

int main (int argc, char *argv[])
{
    print_primes(get_primes(500));

    return 0;
}

File: src/numeric/primes500/primes.lua

local primes = coroutine.wrap(function()
    coroutine.yield(2)
    local cache = {2}
    local candidate = 3
    while true do
        local isprime = true
        for i, prime in ipairs(cache) do
            if candidate % prime == 0 then
                isprime = false
                break
            end
        end
        if isprime then
            coroutine.yield(candidate)
            table.insert(cache, candidate)
        end

        candidate = candidate + 2
    end
end)

function take(n, coroutine)
    local results = {}
    local count = 1
    for value in coroutine do
        results[count] = value
        count = count + 1
        if count > n then
            break
        end
    end
    return results
end

function print_primes(primes)
    print("First Five Hundred Primes")
    for i = 1, 50 do
        print(string.format("     %04d %04d %04d %04d %04d %04d %04d %04d %04d %04d",
            primes[i], primes[i + 50], primes[i + 100], primes[i + 150],
            primes[i + 200], primes[i + 250], primes[i + 300], primes[i + 350],
            primes[i + 400], primes[i + 450]))
    end
end

print_primes(take(500, primes))
